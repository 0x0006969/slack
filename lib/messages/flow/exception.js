const supportLink = require('../../support-link');

module.exports = class Exception {
  constructor(errorCode) {
    this.link = supportLink({
      comments: `Tell us a little more about what happened and include the error code below.\n\nError code: ${errorCode}`,
    });
  }

  toJSON() {
    return {
      response_type: 'ephemeral',
      attachments: [
        {
          text: 'Sorry, we had trouble with your request, but we\'re looking into it! ' +
            `If this is interfering with your work, please <${this.link}|get in touch>!`,
          color: 'danger',
          mrkdwn_in: ['text'],
        },
      ],
    };
  }
};
