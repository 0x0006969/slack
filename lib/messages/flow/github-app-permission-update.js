const ErrorMessage = require('./error-message');

module.exports = class GitHubAppPermissionUpdate extends ErrorMessage {
  constructor({ installation }) {
    super();
    this.installation = installation;
  }
  get guidance() {
    if (this.owner.id === parseInt(this.gitHubUser.id, 10)) {
      return '';
    }
    if (this.owner.type === 'User') {
      return '\n_Note: You will need to ask the owner of the repository to install it for you. ' +
      `Give them <${this.installLink}|this link.>_`;
    }
    return '\n_Note: You need to be an organization owner to install the app (or ask one to install it for you)._';
  }

  toJSON() {
    const message = this.getErrorMessage();
    message.attachments[0].text = 'This feature requires updated permissions.';
    message.attachments[0].actions = [{
      style: 'primary',
      text: 'Update GitHub App permissions',
      type: 'button',
      url: `${this.installation.html_url}/permissions/update`,
    }];
    return message;
  }
};
