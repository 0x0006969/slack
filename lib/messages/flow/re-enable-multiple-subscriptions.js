const { Message } = require('..');

module.exports = class ReEnableMultipleSubscriptions extends Message {
  constructor(subscribedRepositories, creator, reason) {
    super({});
    this.plural = subscribedRepositories.length !== 1;
    this.repositories = subscribedRepositories;
    this.creator = creator;
    this.reason = reason;
  }

  get introduction() {
    return `${this.plural ? 'Subscriptions' : 'The subscription'} to ` +
     `${this.repositories.length} ${this.plural ? 'repositories have' : 'repository has'} been disabled, ` +
      `because <@${this.creator}>, who originally set ${this.plural ? 'them' : 'it'} up, has ${this.reason}.\n` +
      `Use the following slash command${this.plural ? 's' : ''} to re-enable the subscription${this.plural ? 's' : ''}:`;
  }

  get commands() {
    return this.repositories.map(repository => (
      `/github subscribe ${repository.full_name}`
    )).join('\n');
  }

  getAttachment() {
    return {
      ...this.getBaseMessage(),
      mrkdwn_in: ['text'],
      text: `${this.introduction}\n${this.commands}`,
    };
  }

  toJSON() {
    return {
      attachments: [this.getAttachment()],
      response_type: 'in_channel',
    };
  }
};
