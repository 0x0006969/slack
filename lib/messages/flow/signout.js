const { Message } = require('..');

module.exports = class SignOut extends Message {
  constructor(signInLink, userSlackId) {
    super({});
    this.signInLink = signInLink;
    this.userSlackId = userSlackId;
  }

  toJSON() {
    return {
      response_type: 'ephemeral',
      attachments: [{
        ...this.getBaseMessage(),
        fallback: 'Your GitHub account has been disconnected.',
        text: `:white_check_mark: <@${this.userSlackId}> is disconnected from your GitHub account\n` +
          'Features like subscriptions and unfurls will stop working. ' +
          'To re-connect your GitHub account click the button below.',
        actions: [{
          type: 'button',
          text: 'Connect GitHub account',
          url: this.signInLink,
          style: 'primary',
        }],
      }],
    };
  }
};
