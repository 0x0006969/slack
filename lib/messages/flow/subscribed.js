const { Message } = require('..');

module.exports = class Subscribed extends Message {
  constructor({ fromRepository, unsubscribed }) {
    super({});
    this.fromRepository = fromRepository;
    this.unsubscribed = unsubscribed;
  }

  toJSON() {
    const predicate = this.unsubscribed ? 'Unsubscribed' : 'Subscribed';
    const preposition = this.unsubscribed ? 'from' : 'to';
    return {
      response_type: 'in_channel',
      attachments: [{
        ...this.getBaseMessage(),
        text: `${predicate} ${preposition} <${this.fromRepository.html_url}|${this.fromRepository.full_name}>`,
      }],
    };
  }
};
