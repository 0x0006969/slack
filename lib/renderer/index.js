/*
- with unfurls we only have one attachment available to us, in all other cases we have multiple
- take as input the type of message. (e.g. Issue, unfurl; pr, not unfurl)
- based on type, decide capabilities: render_timestamp=true, etc.
*/
const moment = require('moment');

const { getHexColorbyState, arrayToFormattedString } = require('.././helpers');

class Issue {
  constructor(
    issue,
    repository,
    eventType, // e.g. issues.opened
    unfurl = false,
  ) {
    this.capabilities = {
      hasTimestamp: true,
      hasAuthor: true,
      hasOpenState: true,
      canHaveStateMerged: false,
    };

    this.issue = issue;
    this.createdAt = moment(issue.created_at);
    this.repository = repository;

    this.eventType = eventType;
    this.unfurl = unfurl;

    this.major = {
      'issues.opened': true,
    };
  }
  getColor() {
    return getHexColorbyState(this.issue.state);
  }
  getAuthor() {
    return {
      author_name: this.issue.user.login,
      author_link: this.issue.user.html_url,
      author_icon: this.issue.user.avatar_url,
    };
  }
  getCore() {
    // TODO: Need to convert markdown in body to Slack markdown
    const text = this.issue.body;
    const title = `#${this.issue.number} ${this.issue.title}`;
    // eslint-disable-next-line camelcase
    const title_link = this.issue.html_url;
    if (this.unfurl) {
      return {
        text,
        title,
        title_link,
        fallback: title,
      };
    }
    const core = {
      title,
      title_link,
    };
    if (this.major[this.eventType]) {
      core.text = text;
    }
    const preText = action => `[${this.repository.full_name}] Issue ${action} by ${this.issue.user.login}`;
    // TODO: refactor and match on regex
    if (this.eventType === 'issues.opened') {
      core.pretext = preText('opened');
      core.fallback = core.pretext;
    } else if (this.eventType === 'issues.closed') {
      core.pretext = preText('closed');
      core.fallback = core.pretext;
    } else if (this.eventType === 'issues.reopened') {
      core.pretext = preText('reopened');
      core.fallback = core.pretext;
    }
    return core;
  }
  getFields() {
    // projects should be a field as well, but seems to not be easily available via API?
    if (!this.major[this.eventType] && !this.unfurl) {
      return null;
    }
    const fields = [
      {
        title: 'Assignees',
        value: arrayToFormattedString(this.issue.assignees, 'login'),
      },
      {
        title: 'Labels',
        value: arrayToFormattedString(this.issue.labels, 'name'),
      },
      {
        title: 'Comments',
        value: this.issue.comments,
      },
      {
        title: 'Milestone',
        value: this.issue.milestone,
      },
    ];
    return fields.filter(field => field.value)
      .map(field => ({ ...field, short: true }))
      .slice(0, 2);
  }
  getFooter() {
    return {
      footer: `<${this.issue.html_url}|View it on GitHub>`,
      footer_icon: 'https://assets-cdn.github.com/favicon.ico',
    };
  }
  getFullSlackMessageAttachment() {
    const attachment = {
      color: this.getColor(),
      ...this.getAuthor(),
      ...this.getCore(),
      ts: this.createdAt.unix(),
      mrkdwn_in: ['text'],
      ...this.getFooter(),
    };
    // Ideally the below logic should apply to all keys
    // in the attachment object, but not sure what the
    // most beautiful way to do that is.
    if (this.getFields()) {
      attachment.fields = this.getFields();
    }
    return attachment;
  }
  getRenderedMessage() {
    return {
      attachments: [
        this.getFullSlackMessageAttachment(),
      ],
    };
  }
}

class Message {
  constructor(GitHubObjectType, unfurl = false) {
    this.gitHubObjectType = GitHubObjectType;
    this.unfurl = unfurl;
  }

  renderAndGet() {
    // need seperate logic in here for single unfurl attachment and multiple attachments
    return {
      attachments: [
        this.attachment,
      ],
    };
  }

}

// TODO: /github test-run -> delivers all webhooks we're currently ready to receive

module.exports = {
  Message,
  Issue,
};
