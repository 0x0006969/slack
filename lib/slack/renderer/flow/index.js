/* eslint-disable global-require */
const Subscribed = require('./subscribed');
const InvalidUrl = require('./invalid-url');
const InstallGitHubApp = require('./install-github-app');
const NotFound = require('./not-found');
const SignIn = require('./signin');
const Help = require('./help');

module.exports = {
  InvalidUrl,
  InstallGitHubApp,
  NotFound,
  Subscribed,
  SignIn,
  Help,
};
