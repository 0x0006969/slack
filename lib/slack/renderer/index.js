const constants = {
  GITHUB_BLACK: '#24292f',
  CLOSED_RED: '#cb2431',
  OPEN_GREEN: '#36a64f',
  MERGED_PURPLE: '#6f42c1',
  STATUS_SUCCESS: '#28a745',
  STATUS_PENDING: '#dbab09',
  STATUS_FAILURE: '#cb2431',
  BASE_ATTACHMENT_COLOR: '#24292f',
  ATTACHMENT_FIELD_LIMIT: 2,
  MAJOR_MESSAGES: {
    'pull_request.opened': true,
    'issues.opened': true,
  },
};

class Message {
  constructor(constructorObject) {
    this.footerURL = constructorObject.footerURL;
    this.footer = constructorObject.footer;
    this.includeAuthor = constructorObject.includeAuthor;
    this.author = constructorObject.author;
  }

  static cleanFields(
    fields,
    fieldLimit = constants.ATTACHMENT_FIELD_LIMIT,
    short = true,
  ) {
    return fields
      .filter(field => field.value)
      .map((field) => {
        if (short) {
          return { ...field, short: true };
        }
        return { ...field };
      })
      .slice(0, fieldLimit);
  }

  static convertToCondensedAttachment(attachment) {
    const condensedAttachment = {
      ...attachment,
    };
    if (condensedAttachment.fields) {
      condensedAttachment.fields = attachment.fields.slice(0, 2);
    }
    delete condensedAttachment.text;
    return condensedAttachment;
  }

  getBaseMessage() {
    const baseMessage = {
      color: constants.GITHUB_BLACK,
    };
    if (this.footer) {
      baseMessage.footer = this.footer;
      baseMessage.footer_icon = 'https://assets-cdn.github.com/favicon.ico';
    }
    if (this.includeAuthor) {
      baseMessage.author_name = this.author.login;
      baseMessage.author_icon = this.author.avatarURL;
      baseMessage.author_link = this.author.htmlURL;
    }
    return baseMessage;
  }
}

// TODO: /github test-run -> delivers all webhooks we're currently ready to receive

module.exports = {
  Message,
  constants,
};
