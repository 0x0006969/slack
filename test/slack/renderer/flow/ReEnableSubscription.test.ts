import { ReEnableSubscription } from "../../../../lib/slack/renderer/flow";

describe("ReEnableSubscription message rendering", () => {
  test("works", () => {
    const message = new ReEnableSubscription("atom/atom", "U012345");
    expect(message.toJSON()).toMatchSnapshot();
  });
});
