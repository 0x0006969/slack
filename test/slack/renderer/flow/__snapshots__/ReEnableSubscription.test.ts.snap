// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReEnableSubscription message rendering works 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "The subscription to \`atom/atom\` has been disabled, because the subscription creator (<@U012345>) no longer has access.
Run \`/github subscribe atom/atom\` to re-enable the subscription",
    },
  ],
  "response_type": "in_channel",
}
`;
